### PR TITLE
Implement and use registerIreeDialects()

### PR DESCRIPTION
**Pull request by marbre**
* Equivalent to `registerMlirDialects()` for IREE dialects
* Allows to avoid alwayslink

### CODE REVIEW

**Comment by marbre** [comment]
I am currently struggling with the Bazel build which fails when I remove the `alwayslink` flags from the config. Must be something I am not seeing. Unfortunately, I don't have local build environment with Bazel. Would it be possible that somebody takes a look?

**Comment by GMNGeoffrey** [comment]
Yeah so looks like we can remove alwayslink from every BUILD file except for iree/compiler/Dialect/IREE/IR/BUILD without errors. When I remove it from that it can't find the things in that dialect because that dialect isn't listed in your registration function :-) I added it and everything passes fine: https://github.com/GMNGeoffrey/iree/commit/40d8c2dc412c78bb0e36474d2994da888caf0085

**Comment by GMNGeoffrey** [comment]
I still can't push to the fork though. I think you don't have "maintainers can edit" on

**Comment by GMNGeoffrey** [comment]
I'll go ahead and merge as soon as the CI passes :-)

**Comment by marbre** [comment]
> I'll go ahead and merge as soon as the CI passes :-)

Great, many thanks!

**Comment by GMNGeoffrey** [comment]
FYI there are some missing deps I need to add in the build files to fix things up internally

File: iree/tools/BUILD [hunk 1]
```diff
@@ -238,7 +238,6 @@ cc_library(
         "@llvm-project//mlir:Translation",
         "@org_tensorflow//tensorflow/compiler/mlir/xla:xla_dialect_registration",
     ] + IREE_COMPILER_DIALECT_MODULES + IREE_COMPILER_TARGET_BACKENDS,
-    alwayslink = 1,
```
Comment (GMNGeoffrey) [inline iree/tools/BUILD:241]: I think removing alwayslink here isn't right. iree_translate_main has the main function so it needs to always be linked in. This doesn't have to do with dialect registration

Comment (marbre) [reply iree/tools/BUILD:241]: Okay, I'll drop the two commits that touch the iree/tools configs.

**Pull request closed without merging**